### DISCUSSION TITLE
v3 how to livewire publish assets?

### DISCUSSION THREAD

**Discussion by faizananwerali** [livewire/livewire#6575] [1. Help] [answered]
publishing assets in v3 is removed, Why? how can I do it? dynamic assets are not working me, giving me 404 error.

**Comment by PhiloNL** [comment]
It has not been removed:
```shell
php artisan livewire:publish --assets
```

**Reply by faizananwerali** [reply]
then how would you publish? because currently, it doesn't publish livewire.js files.

This is what I have manually done.

In composer.json

```json
        "post-update-cmd": [
            "@php artisan vendor:publish --tag=laravel-assets --ansi --force",
            "@php artisan vendor:publish --tag=livewire:assets --ansi --force",
            "cp vendor/livewire/livewire/dist/livewire.js public/assets/livewire/livewire.js",
            "cp vendor/livewire/livewire/dist/livewire.min.js public/assets/livewire/livewire.min.js",
            "cp vendor/livewire/livewire/dist/livewire.esm.js public/assets/livewire/livewire.esm.js",
            "cp vendor/livewire/livewire/dist/manifest.json public/assets/livewire/manifest.json"
        ],
```

In AppServiceProvider.php

```php
    public function boot(): void
    {
        Livewire::setScriptRoute(function ($handle) {
            return Route::get('/assets/livewire/livewire.js', $handle);
        });  
    }
```

**Reply by PhiloNL** [reply]
Out of curiosity, why are you publishing all these assets?

**Reply by PhiloNL** [reply]
Also, are you using Cloudpanel? If so, [this might be the fix for you](https://github.com/livewire/livewire/discussions/6328#discussioncomment-6845945).

**Reply by N1ebieski** [reply]
@PhiloNL In my opinion, this is not a solution. Static files like .js should be served via nginx, not php-fpm.

It is better to [import Livewire ESM into the main file app.js in vite](https://livewire.laravel.com/docs/upgrading#accessing-alpine-via-js-bundle).

**Reply by soundslocke** [reply]
I concur with @N1ebieski: serving static assets via PHP doesn't make a ton of sense in production. Why invoke FastCGI, the PHP runtime, Composer autoloading overhead, Laravel kernel and facilities, and subsequently the Livewire facilities, simply to serve a static file? This is what nginx, etc, do best.

Livewire v2 supported publishing the assets and I'm confused why v3 does not.

**Reply by faizananwerali** [reply]
@soundslocke yeah, worst decision ever.

**Reply by soundslocke** [reply]
:laughing: Haha, maybe not the "worst decision ever" per se, but it seems like a good idea to reintroduce the built-in command to publish the static assets for ideal serving in production.

**Reply by joshhanley** [reply]
@faizananwerali while it may be frustrating, it's likely to just be a bug, not necessarily a decision to remove it. If we intentionally removed it, it would have been in the upgrade guide. If you have a look the `livewire:publish --assets` command still exists, just isn't working correctly. PRs to fix this issue are welcome 🙂

But I agree with @PhiloNL that importing Livewire assets in your `app.js` is the best solution, as published assets can become out of date, whereas an import will always use the latest version every time you build your project.

**Reply by faizananwerali** [reply]
@joshhanley While you may be right that importing will always use the latest version, I don't see the point of it. There is no difference between importing and publishing assets. Publishing assets command with `post-update-cmd` will also use the latest version and keep the assets file and code separate. I like it that way. It used to work perfectly that way. Other open-source projects used to work it that way. However, Livewire's decision to remove this feature has left me feeling somewhat frustrated. Why remove the perfectly working feature? Why only one way to import?

**Reply by joshhanley** [reply]
@faizananwerali while I understand your frustration, ask that you please read my comment again. We didn't remove anything intentionally, we have done a ground-up rewrite of Livewire V3 so it was likely just missed or has bugs. PRs to fix are welcome.

**Comment by PhiloNL** [comment]
> @PhiloNL In my opinion, this is not a solution. Static files like .js should be served via nginx, not php-fpm.
> 
> It is better to [import Livewire ESM into the main file app.js in vite](https://livewire.laravel.com/docs/upgrading#accessing-alpine-via-js-bundle).

Agree but some people don't want to do this. Hence my question, why publish the assets.

**Comment by faizananwerali** [comment, marked as answer]
Thanks, everyone found my answer here [https://github.com/livewire/livewire/discussions/6328](https://github.com/livewire/livewire/discussions/6328)
